schema_utils.py

```diff
@@ -185,6 +204,224 @@ def get_validator(validator_id):
     return _Validators.get(validator_id)
 
 
+def is_numeric_operand(string):
+    """Checks if the given string represents a valid number.
+
+    Args:
+        string: str. The operand to be validated.
+
+    Returns:
+        bool. Whether the given string is a valid numeric operand.
+    """
+    try:
+        float(string)
+        return True
+    except ValueError:
+        return False
+
+
+def is_algebraic_operand(string):
+    """Checks if the given string represents a valid algebraic operand
+    (variable). Given string must be a singular english alphabet or a valid
+    greek letter.
+
+    Args:
+        string: str. The operand to be validated.
+
+    Returns:
+        bool. Whether the given string is a valid algebraic operand.
+    """
+    return (
+        string.isalpha() and len(string) == 1) or string in GREEK_LETTERS.keys()
+
+
+def is_operand(string):
+    """Checks if the given string represents a valid algebraic/numeric operand.
+
+    Args:
+        string: str. The operand to be validated.
+
+    Returns:
+        bool. Whether the given string is a valid algebraic/numeric operand.
+    """
+    return is_numeric_operand(string) or is_algebraic_operand(string)
+
+
+def contains_balanced_brackets(expression):
+    """Checks if the given expression contains a balanced bracket sequence.
+
+    Args:
+        expression: str. A math expression (algebraic/numeric).
+
+    Returns:
+        bool. Whether the given expression contains a balanced
+            bracket sequence.
+    """
+    openers, closers = '({[', ')}]'
+    stack = []
+    for character in expression:
+        if character in openers:
+            stack.append(character)
+        elif character in closers:
+            if len(stack) == 0:
+                return False
+            top_element = stack.pop()
+            if openers.index(top_element) != closers.index(character):
+                return False
+    return len(stack) == 0
+
+
+def tokenize(expression):
```

Comment:
I just realized something -- every function in this file should have its own unit tests. In particular, you should also have tests for tokenize(). Could you please add them?

(I was trying to figure out how numbers with multiple decimal points in them are handled and AFAICT the tokenize() function seems to just write the whole thing as a token even though it's invalid, but the tests would help verify things like this.)

Reply:
Comment is obsolete due to new changes. These functions don't exist anymore.